### PR TITLE
Python REPL: Persistent Storage, Multi Fabric, JupyterLab Support,

### PR DESCRIPTION
**Pull request by mrjerryjohns**
This adds support for the following features in the Python REPL:

- Added instructions for building the REPL
- Creates a new "REPL Playground", which is a Jupyter Lab instance of the REPL for interactive development
- Adds a number of Jupyter Notebooks that serve as both guides on using the various features of the REPL as well as being runnable within the playground interactively to streamline usage of its various features
- Persistent storage support in the form of a JSON configuration file whose path can be over-riden at REPL initialization. JSON was chosen to make it easy to debug and read the configuration file.
- Creation of a new `FabricAdmin` class that manages the fabric, while `ChipDeviceController` shifts away from being a singleton to encapsulating a controller instance on that fabric. We can now instantiate multiple fabrics in the REPL as well as commission a device onto multiple such fabrics.
- Fix-ups to the chip-device-ctrl to work with this new model.

Testing:
- Validated the various Notebooks to ensure all existing and newly added
  functionality work well. This includes commissioning a target onto
  multiple fabrics and ensuring IM interactions work.

### CODE REVIEW

File: scripts/build_python.sh [hunk 1]
```diff
@@ -105,13 +105,15 @@ source "$CHIP_ROOT/scripts/activate.sh"
 [[ -n "$chip_case_retry_delta" ]] && chip_case_retry_arg="chip_case_retry_delta=$chip_case_retry_delta" || chip_case_retry_arg=""
 
 gn --root="$CHIP_ROOT" gen "$OUTPUT_ROOT" --args="chip_detail_logging=$chip_detail_logging enable_pylib=$enable_pybindings enable_rtti=$enable_pybindings chip_project_config_include_dirs=[\"//config/python\"] $chip_mdns_arg $chip_case_retry_arg"
+echo gn --root="$CHIP_ROOT" gen "$OUTPUT_ROOT" --args="chip_detail_logging=$chip_detail_logging enable_pylib=$enable_pybindings enable_rtti=$enable_pybindings chip_project_config_include_dirs=[\"//config/python\"] $chip_mdns_arg $chip_case_retry_arg"
```
Comment (andy31415) [inline scripts/build_python.sh:108]: do we need this?
Could we maybe do a `set -x` in a sub-shell if we need printout so that we do not rely on perfect copy and paste?

File: scripts/build_python.sh [hunk 2]
```diff
@@ -105,13 +105,15 @@ source "$CHIP_ROOT/scripts/activate.sh"
...
 
 # Compiles python files
 # Check pybindings was requested
 if [ "$enable_pybindings" == true ]; then
     ninja -C "$OUTPUT_ROOT" pycontroller
 else
     ninja -C "$OUTPUT_ROOT" python
+    echo ninja -C "$OUTPUT_ROOT" python
```
Comment (andy31415) [inline scripts/build_python.sh:116]: same here. Generally wondering if we should move this into build_examples since that one has dry run to show actual used commands (but I explicitly am not asking for this PR  to use build_examples ... that is a reaasonably sized PR by itself)

File: src/controller/CHIPDeviceController.cpp [hunk 3]
```diff
@@ -339,6 +339,7 @@ void DeviceController::PersistNextKeyId()
     if (mStorageDelegate != nullptr && mState == State::Initialized)
     {
         uint16_t nextKeyID = mIDAllocator.Peek();
+        printf("PersistNextKey %d\n", nextKeyID);
```
Comment (andy31415) [inline src/controller/CHIPDeviceController.cpp:342]: Delete debug code.

File: src/controller/CHIPDeviceController.cpp [hunk 4]
```diff
@@ -619,6 +620,7 @@ CHIP_ERROR DeviceCommissioner::Init(CommissionerInitParams params)
 
     params.systemState->SessionMgr()->RegisterRecoveryDelegate(*this);
 
+#if 0
```
Comment (andy31415) [inline src/controller/CHIPDeviceController.cpp:623]: Is this intentional?

File: src/controller/python/OpCredsBinding.cpp [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Implementation of the native methods expected by the Python
+ *      version of Chip Device Manager.
```
Comment (andy31415) [inline src/controller/python/OpCredsBinding.cpp:24]: Lets delete the doc comments for files. They are generally not helping reviews and often copy and pasted so wrong (here the OpCreds is probably not device manager)

File: src/controller/python/chip-device-ctrl.py [hunk 6]
```diff
@@ -299,6 +328,8 @@ def do_setlogoutput(self, line):
         Set the level of Chip logging output.
         """
 
+        # warnings.warn("This method is being deprecated. Please use the DeviceController.SetLogFilter method directly in the REPL", DeprecationWarning)
```
Comment (andy31415) [inline src/controller/python/chip-device-ctrl.py:331]: Do we want commented out code for future deprecation?
Consider a helper like `def warnDeprecated(self, alternative=None)` to have shorter lines.

File: src/controller/python/chip/FabricAdmin.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    activeFabricIndexList = set()
+    activeFabricIdList = set()
+
+    activeAdmins = set()
+
+    def AllocateNextFabricIndex(self):
+        ''' Allocate the next un-used fabric index.
+        '''
+        nextFabricIndex = 1
+        while (True):
```
Comment (andy31415) [inline src/controller/python/chip/FabricAdmin.py:77]: is this:

```python
while nextFabricIndex in FabricAdmin.activeFabricIndexList:
    nextFabricIndex = nextFabricIndex + 1
return nextFabricIndex
```

?

Comment (andy31415) [reply src/controller/python/chip/FabricAdmin.py:77]: same for below.

File: src/protocols/secure_channel/SessionIDAllocator.h [hunk 8]
```diff
@@ -46,7 +46,7 @@ class SessionIDAllocator
     static constexpr uint16_t kMaxSessionID       = UINT16_MAX;
     static constexpr uint16_t kUnsecuredSessionId = 0;
 
-    uint16_t mNextAvailable = 1;
+    static uint16_t mNextAvailable;
```
Comment (andy31415) [inline src/protocols/secure_channel/SessionIDAllocator.h:49]: Is this intentional? 
We may want a comment on why static is correct and desirable here.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
It looks like there are pieces of other PRs in here too... Please merge those and do a clean PR here?

**Comment by mrjerryjohns** [comment]
> It looks like there are pieces of other PRs in here too... Please merge those and do a clean PR here?

Yes, there are. This PR is just put out to get reviews started, but it has bits from #13569 as well. I'm trying to get that merged today.

File: src/controller/python/test/test_scripts/base.py [hunk 9]
```diff
@@ -133,13 +137,54 @@ def TestDiscovery(self, discriminator: int):
...
+        chip.FabricAdmin.FabricAdmin.ShutdownAll()
+
+        self.fabricAdmin = chip.FabricAdmin.FabricAdmin(
+            fabricId=1, fabricIndex=1)
+        fabricAdmin2 = chip.FabricAdmin.FabricAdmin(fabricId=2, fabricIndex=2)
+
+        self.devCtrl = self.fabricAdmin.NewController(self.controllerNodeId)
+        devCtrl2 = fabricAdmin2.NewController(self.controllerNodeId)
+
+        await self.devCtrl.ReadAttribute(nodeid, [(Clusters.Basic)])
```
Comment (erjiaqing) [inline src/controller/python/test/test_scripts/base.py:180]: Can we read some fabric scoped attributes and verify the results? For example, try to read `Clusters.OperationalCredentials.Attributes.CurrentFabricIndex` instead of `Clusters.Basic` here?

File: src/controller/python/chip/storage/__init__.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+_SyncSetKeyValueCbFunct = CFUNCTYPE(
+    None, py_object, c_char_p, POINTER(c_char),  c_uint16)
+_SyncGetKeyValueCbFunct = CFUNCTYPE(
+    None, py_object, c_char_p, POINTER(c_char), POINTER(c_uint16))
+_SyncDeleteKeyValueCbFunct = CFUNCTYPE(None, py_object, POINTER(c_char_p))
+
+
+@_SyncSetKeyValueCbFunct
+def _OnSyncSetKeyValueCb(storageObj, key: str, value, size):
+    storageObj.SetSdkKey(key.decode("utf-8"), ctypes.string_at(value, size))
```
Comment (erjiaqing) [inline src/controller/python/chip/storage/__init__.py:44]: calling decode() on `str` type seems not correct, c_char_p is mapped to bytes actually

File: src/controller/python/chip/storage/__init__.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+            count = count + 1
+
+        size[0] = count
+    else:
+        size[0] = 0
+
+
+@_SyncDeleteKeyValueCbFunct
+def _OnSyncDeleteKeyValueCb(storageObj, key):
+    print("OnSyncDeleteKeyValue")
```
Comment (erjiaqing) [inline src/controller/python/chip/storage/__init__.py:72]: Using logging or remove this log?

File: src/controller/python/chip/storage/__init__.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    storageObj.SetKey(key, None)
```
Comment (erjiaqing) [inline src/controller/python/chip/storage/__init__.py:73]: PersistentStorage does not has `SetKey`, seems it should be `SetSdkKey`?
Also, the type of key is bytes instead of str

File: src/controller/python/ChipDeviceController-ScriptBinding.cpp [hunk 13]
```diff
@@ -199,10 +265,13 @@ ChipError::StorageType pychip_DeviceController_NewDeviceController(chip::Control
     const chip::Credentials::AttestationTrustStore * testingRootStore = chip::Credentials::GetTestAttestationTrustStore();
     SetDeviceAttestationVerifier(GetDefaultDACVerifier(testingRootStore));
 
-    CHIP_ERROR err = sOperationalCredentialsIssuer.Initialize(sStorageDelegate);
-    VerifyOrReturnError(err == CHIP_NO_ERROR, err.AsInteger());
+    chip::Controller::ExampleOperationalCredentialsIssuer * operationalCredentialsIssuer =
```
Comment (erjiaqing) [inline src/controller/python/ChipDeviceController-ScriptBinding.cpp:268]: I did not see any place freeing this pointer

Possible solution is to create a new class inherits from chip::Controller::DeviceCommissioner with this member and make this function a wrapper of chip::Controller::Python::DeviceCommissioner() (or Init, or something else)

Maybe a TODO here is OK?

File: src/controller/python/ChipDeviceController-ScriptBinding.cpp [hunk 14]
```diff
@@ -181,14 +183,78 @@ uint64_t pychip_GetCommandSenderHandle(chip::DeviceProxy * device);
...
+
+    return CHIP_NO_ERROR.AsInteger();
 }
 
 ChipError::StorageType pychip_DeviceController_NewDeviceController(chip::Controller::DeviceCommissioner ** outDevCtrl,
                                                                    chip::NodeId localDeviceId)
 {
+    static uint16_t sFabricIndex = 1;
+
     *outDevCtrl = new chip::Controller::DeviceCommissioner();
```
Comment (erjiaqing) [inline src/controller/python/ChipDeviceController-ScriptBinding.cpp:253]: Realized any error below could lead to memory leak, I guess this could be resolved with the comment below in another PR?

File: src/controller/python/OpCredsBinding.cpp [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+
+ChipError::StorageType pychip_OpCreds_AllocateController(OpCredsContext * context,
+                                                         chip::Controller::DeviceCommissioner ** outDevCtrl, uint8_t fabricIndex,
+                                                         FabricId fabricId, chip::NodeId nodeId)
+{
+    ChipLogDetail(Controller, "Creating New Device Controller");
+
+    VerifyOrReturnError(context != nullptr, CHIP_ERROR_INVALID_ARGUMENT.AsInteger());
+
+    *outDevCtrl = new chip::Controller::DeviceCommissioner();
```
Comment (erjiaqing) [inline src/controller/python/OpCredsBinding.cpp:132]: Suggest using unique_ptr and call release() before return.

**Comment by woody-apple** [comment]
Fast tracking, given > 3 business days for review.

File: scripts/build_python.sh (hunk 1 above)
Comment (mrjerryjohns) [reply scripts/build_python.sh:108]: For now, I'll just delete it.

File: scripts/build_python.sh (hunk 2 above)
Comment (mrjerryjohns) [reply scripts/build_python.sh:116]: For now, I'll just delete it.

File: src/controller/python/chip-device-ctrl.py (hunk 6 above)
Comment (mrjerryjohns) [reply src/controller/python/chip-device-ctrl.py:331]: I'll need to add a bunch more in a future PR, so will re-factor then.

File: src/protocols/secure_channel/SessionIDAllocator.h (hunk 8 above)
Comment (mrjerryjohns) [reply src/protocols/secure_channel/SessionIDAllocator.h]: This was merged in a separate PR, so resolved.

File: src/controller/python/chip/storage/__init__.py (hunk 10 above)
Comment (mrjerryjohns) [reply src/controller/python/chip/storage/__init__.py:44]: It works however, since I've validated setting and getting keys as part of normal operation.

Comment (mrjerryjohns) [reply src/controller/python/chip/storage/__init__.py:44]: What should it be?

File: src/controller/python/ChipDeviceController-ScriptBinding.cpp (hunk 13 above)
Comment (mrjerryjohns) [reply src/controller/python/ChipDeviceController-ScriptBinding.cpp:268]: Good catch. I'll free it in `pychip_DeviceController_DeleteDeviceController`, which is the symmetric equivalent here.

File: src/controller/python/chip/storage/__init__.py (hunk 10 above)
Comment (erjiaqing) [reply src/controller/python/chip/storage/__init__.py:44]:
```diff
- def _OnSyncSetKeyValueCb(storageObj, key: str, value, size):
+ def _OnSyncSetKeyValueCb(storageObj, key: bytes, value, size):
```

File: src/controller/CHIPDeviceController.cpp [hunk 16]
```diff
@@ -1444,8 +1449,9 @@ void DeviceCommissioner::OnSessionEstablishmentTimeout()
 
 void DeviceCommissioner::OnSessionEstablishmentTimeoutCallback(System::Layer * aLayer, void * aAppState)
 {
-    static_cast<DeviceCommissioner *>(aAppState)->OnSessionEstablishmentTimeout();
+    static_cast<DeviceCommissioner *>(aAppState)->OnSessionEstablishmentError(CHIP_ERROR_TIMEOUT);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1452]: Why this change?

File: src/controller/CHIPDeviceController.cpp [hunk 17]
```diff
@@ -928,6 +930,9 @@ CHIP_ERROR DeviceCommissioner::Commission(NodeId remoteDeviceId, CommissioningPa
         ChipLogError(Controller, "Invalid device for commissioning " ChipLogFormatX64, ChipLogValueX64(remoteDeviceId));
         return CHIP_ERROR_INCORRECT_STATE;
     }
+
+    mCommissioningStage = kSecurePairing;
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:934]: Why this change?

File: src/controller/CHIPDeviceControllerFactory.h [hunk 18]
```diff
@@ -106,6 +106,13 @@ class DeviceControllerFactory
     DeviceControllerFactory(DeviceControllerFactory const &) = delete;
     void operator=(DeviceControllerFactory const &) = delete;
 
+    //
+    // Some clients do not prefer a complete shutdown of the stack being initiated if
+    // all device controllers have ceased to exist. To avoid that, this method has been
+    // created to permit retention of the underlying system state to avoid that.
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceControllerFactory.h:112]: Who will release it and when?

File: src/controller/ExampleOperationalCredentialsIssuer.h [hunk 19]
```diff
@@ -42,6 +42,16 @@ namespace Controller {
...
 {
 public:
+    //
+    // Constructor to create an instance of this object that vends out operational credentials for a given fabric.
+    //
+    // An index should be provided to numerically identify this instance relative to others in a multi-fabric deployment. This is
+    // needed given the interactions of this object with persistent storage. Consequently, the index is used to scope the entries
+    // read/written to/from storage.
+    //
+    // It is recommended that this index track the fabric index within which this issuer is operating in.
```
Comment (bzbarsky-apple) [inline src/controller/ExampleOperationalCredentialsIssuer.h:52]:
```suggestion
    // It is recommended that this index track the fabric index within which this issuer is operating.
```

File: src/controller/CHIPDeviceController.cpp (hunk 17 above)
Comment (mrjerryjohns) [reply src/controller/CHIPDeviceController.cpp:934]: Stale change from way back to fix a bug in this state machine that has since been fixed.

File: src/controller/CHIPDeviceController.cpp (hunk 16 above)
Comment (mrjerryjohns) [reply src/controller/CHIPDeviceController.cpp:1452]: I had a good reason for this a few weeks ago, but I can't remember why now....If I can't remember, I'll delete it for now.

File: src/controller/CHIPDeviceControllerFactory.h (hunk 18 above)
Comment (mrjerryjohns) [reply src/controller/CHIPDeviceControllerFactory.h:112]: It'll be released at stack shutdown in the REPL. Have updated the comments and added a release method.
